repo2docker/buildpacks/pipfile/__init__.py

```diff
@@ -0,0 +1,157 @@
+
+        py_version = None
+        for file in files_to_search_in_order:
+            try:
+                with open(file["path"]) as f:
+                    for line in f:
+                        match = re.match(file["pattern"], line)
+                        if not match:
+                            continue
+                        py_version = match.group(1)
+                        break
+            except FileNotFoundError:
+                pass
+            if py_version:
+                break
+
+        # extract major.minor
+        if py_version:
+            if len(py_version) == 1:
+                self._python_version = self.major_pythons.get(py_version[0])
+            else:
+                # return major.minor
+                self._python_version = ".".join(py_version.split(".")[:2])
+            return self._python_version
+        else:
+            # use the default Python
+            self._python_version = self.major_pythons["3"]
+            return self._python_version
+
+    def get_assemble_scripts(self):
+        """Return series of build-steps specific to this repository.
+        """
+        # If we have either Pipfile.lock, Pipfile, or runtime.txt declare the
+        # use of Python 2, Python 2.7 will be made available in the *kernel*
+        # environment. The notebook servers environment on the other hand
+        # requires Python 3 but may require something additional installed in it
+        # still such as `nbgitpuller`. For this purpose, a "requirements3.txt"
+        # file will be used to install dependencies for the notebook servers
+        # environment, if Python 2 had been specified for the kernel
+        # environment.
+        assemble_scripts = super().get_assemble_scripts()
+
+        if self.py2:
+            # using Python 2 as a kernel, but Python 3 for the notebook server
+
+            # requirements3.txt allows for packages to be installed to the
+            # notebook servers Python environment
+            nb_requirements_file = self.binder_path("requirements3.txt")
+            if os.path.exists(nb_requirements_file):
+                assemble_scripts.append(
+                    (
+                        "${NB_USER}",
+                        'pip install --no-cache-dir -r "{}"'.format(
+                            nb_requirements_file
+                        ),
+                    )
+                )
+
+        pipfile = self.binder_path("Pipfile")
+        pipfile_lock = self.binder_path("Pipfile.lock")
+
+        # A Pipfile(.lock) can contain relative references, so we need to be
+        # mindful about where we invoke pipenv as that will dictate where .`
+        # referes to.
+        #     [packages]
+        #     my_package_example = {path=".", editable=true}
+        working_directory = self.binder_dir or "."
+
+        # install pipenv to install dependencies within Pipfile.lock or Pipfile
+        assemble_scripts.append(("${NB_USER}", "pip install pipenv"))
```

Comment:
should we pin the version of pipenv used? Unless there is a good reason not to, I'd vote to pin it

Reply:
Sure, so we pin it the major and minor, but leaves the patch out? like `pipenv=X.Y.*` ?

Reply:
oh... their versions are like this: 2018.11.26

I pinned it to that.